File: tee/cli/common/descriptor.py
"""Node descriptors: the handoff from provisioning to configuration.

The descriptor file is the cohort's `nodes` map — exactly what the
seismic_node Pulumi program's `pulumi stack output nodes --json` prints:

    {
      "devnet-3-1": {"public_ip": "203.0.113.7", "fqdn": "n1.seismicdev.net"},
      "devnet-3-2": {"public_ip": "203.0.113.8", "fqdn": "n2.seismicdev.net"}
    }

Each entry describes one provisioned node: the public IP to reach it at
(`public_ip`) and the FQDN clients use (`fqdn`). The key is the node's
name — the name its harvest record and founding validator slot carry — so
identity lives in the file's content, never in how a file was saved. The
file is the boundary between the infrastructure layer (provisioning, owned
by Pulumi and run standalone) and both CLIs: a CLI consumes the map and
never shells out to or wraps Pulumi. Saving the stack output as-is is one
idempotent command with no per-node bookkeeping, so re-running it after any
map edit (add, remove, re-image) cannot leave the file drifted from the
stack.

A network directory keeps it at `<network>/nodes/nodes.json`, beside
`nodes/bootnodes.json`; a bring-your-own-infra operator (Terraform, manual
console, …) hand-writes a one-key map in the same shape, since only
`public_ip`/`fqdn` are read and any extra keys are ignored:

    seismic-tee-node configure --node my-node.json \
        --bootnode enode://… --manifest m.json
"""

import argparse
import json
from dataclasses import dataclass
from pathlib import Path

# The descriptor map lives under this subdir of a network directory, beside
# the founding bootnode set. Mutable infra state — regenerated per deploy,
# gone with the stack — so it stays gitignored (tee/networks/.gitignore
# `*/nodes/`) while the artifact set around it commits.
NODES_DIRNAME = "nodes"
NODES_FILENAME = "nodes.json"


@dataclass(frozen=True)
class NodeDescriptor:
    """One provisioned node, as the `nodes` map describes it."""

    name: str
    public_ip: str
    fqdn: str


def nodes_file(network_dir: Path) -> Path:
    """Where a network directory keeps its descriptor map."""
    return network_dir / NODES_DIRNAME / NODES_FILENAME


def load_descriptors(path: Path) -> dict[str, NodeDescriptor]:
    """Load and validate a descriptor map, in sorted node-name order.

    `pulumi stack output --json` sorts keys, so a saved stack output is
    already in this order; sorting here means a hand-written map pairs with
    the authored withdrawal credentials the same way. Raises ValueError
    (naming the file and the entry) on any shape the CLIs can't act on:
    a top level that isn't an object, an entry that isn't an object, an
    empty map, or an entry whose `public_ip`/`fqdn` is absent, `null`, or
    empty — reported apart, because the fix differs (see `_require`).
    """
    with open(path) as f:
        data = json.load(f)
    if not isinstance(data, dict):
        raise ValueError(
            f"descriptor file {path} must be a JSON object mapping node name "
            f"→ {{public_ip, fqdn}} (the stack's `nodes` output); got "
            f"{type(data).__name__}"
        )
    if not data:
        raise ValueError(
            f"descriptor file {path} holds no nodes — is the stack's `nodes` map empty?"
        )
    if {"public_ip", "fqdn"} <= set(data) and all(
        not isinstance(v, dict) for v in data.values()
    ):
        # One node's `{public_ip, fqdn}` at the top level — the shape a
        # descriptor file had before it was the map. Say so rather than
        # reporting "node 'fqdn' must be an object".
        raise ValueError(
            f"descriptor file {path} is a single node's {{public_ip, fqdn}}, not "
            "the map; the descriptor file is {<name>: {public_ip, fqdn}, …} — "
            "wrap it under the node's name"
        )
    descriptors = {}
    for name in sorted(data):
        entry = data[name]
        if not isinstance(entry, dict):
            raise ValueError(
                f"descriptor file {path}: node {name!r} must be an object with "
                f"public_ip and fqdn; got {type(entry).__name__}"
            )
        descriptors[name] = NodeDescriptor(
            name=name,
            public_ip=_require(entry, "public_ip", name, path),
            fqdn=_require(entry, "fqdn", name, path),
        )
    return descriptors


def select_descriptor(
    descriptors: dict[str, NodeDescriptor], name: str | None, path: Path
) -> NodeDescriptor:
    """Pick the one node a single-node command acts on.

    With one entry in the file it is the node; with several, `name` says
    which. Raises ValueError naming the file and the choices otherwise.
    """
    if name is None:
        if len(descriptors) == 1:
            return next(iter(descriptors.values()))
        raise ValueError(
            f"descriptor file {path} holds {len(descriptors)} nodes "
            f"({', '.join(descriptors)}); pass --name to say which"
        )
    try:
        return descriptors[name]
    except KeyError:
        raise ValueError(
            f"descriptor file {path} has no node {name!r}; it holds "
            f"{', '.join(descriptors)}"
        ) from None


def add_node_args(parser: argparse.ArgumentParser) -> None:
    """`--node <file> [--name <name>]`: how a single-node command names its
    node. Shared by `seismic-tee-node configure` / `verify` / `status`."""
    parser.add_argument(
        "--node",
        type=Path,
        required=True,
        metavar="FILE",
        help=(
            "Descriptor map JSON: `pulumi stack output nodes --json`, i.e. "
            "{<name>: {public_ip, fqdn}, …} (a network directory keeps it at "
            f"{NODES_DIRNAME}/{NODES_FILENAME}). Provides the node's "
            "public_ip/fqdn. With one entry it is the node; with several, "
            "--name says which."
        ),
    )
    parser.add_argument(
        "--name",
        default=None,
        metavar="NAME",
        help="Which node in --node to act on (its key). Optional when the "
        "file holds exactly one.",
    )


def load_node_arg(args: argparse.Namespace) -> NodeDescriptor:
    """Resolve `--node`/`--name` to the one node, exiting with the reason if
    the file is absent, malformed, or doesn't single out a node."""
    if not args.node.is_file():
        raise SystemExit(f"--node descriptor file not found: {args.node}")
    try:
        return select_descriptor(load_descriptors(args.node), args.name, args.node)
    except json.JSONDecodeError as e:
        raise SystemExit(f"--node {args.node} is not valid JSON: {e}") from None
    except ValueError as e:
        raise SystemExit(str(e)) from None


def _require(entry: dict, key: str, name: str, path: Path) -> str:
    """Return entry[key], or raise naming the file, the node, and the mistake.

    An absent key, an explicit `null`, and an empty string all fail the
    same way — a descriptor that can't reach a node — but the fix differs,
    so they are reported apart: only the absent case is likely a typo and
    helped by listing the keys that *were* there, while a `null` is usually
    Pulumi reporting an output that never got set. Saying "missing" for a
    key the operator can plainly see in the file sends them looking in the
    wrong place.
    """
    if key not in entry:
        raise ValueError(
            f"descriptor file {path}: node {name!r} is missing required key "
            f"{key!r}; got keys {sorted(entry)}"
        )
    value = entry[key]
    if value is None:
        raise ValueError(
            f"descriptor file {path}: node {name!r} has {key!r} set to null "
            "(Pulumi emits null for an output that never got set)"
        )
    if not isinstance(value, str) or not value:
        raise ValueError(
            f"descriptor file {path}: node {name!r} has an empty or non-string "
            f"{key!r}: {value!r}"
        )
    return value

File: tee/cli/rust/common/src/descriptor.rs
//! Node descriptor: the handoff from provisioning to configuration.
//!
//! A *descriptor* is a small JSON file describing one provisioned node — the
//! public IP to reach it at (`public_ip`) and the FQDN clients use (`fqdn`).
//! It is the boundary between the infrastructure layer (provisioning, owned by
//! Pulumi and run standalone) and these CLIs: a CLI consumes a descriptor and
//! never wraps Pulumi.
//!
//! The seismic_node Pulumi program's `nodes` output is one `{public_ip, fqdn}`
//! per node, keyed by name, and one file per node is split out of it. A
//! bring-your-own-infra operator (Terraform, manual console, …) can hand-write
//! the same shape: only those two keys are read, and any extra keys are
//! ignored.
//!
//! The file's stem is the node's name across a cohort
//! (`dev-bootstrap-node-1.json` → `dev-bootstrap-node-1`), which is how the
//! descriptors, the harvest records, and the founding validator slots line up.

use std::path::Path;

use serde::{Deserialize, Deserializer};

use crate::error::{Error, Result};

/// One provisioned node, as named by its descriptor file.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct NodeDescriptor {
    /// The descriptor file's stem.
    pub name: String,
    /// The address the operator-only ports are reached on.
    pub public_ip: String,
    /// The node's DNS name — the cert domain, and where clients reach its RPC.
    pub fqdn: String,
}

/// A key that must carry a usable value, in each of the states it can be in.
///
/// The nesting is what separates an absent key from a present one holding
/// `null`: the outer `Option` is serde's "was the key there at all", the inner
/// one is the `null`. Both are rejected, but they are different mistakes and an
/// operator is told which — a key that isn't there is usually a typo, while a
/// `null` is usually Pulumi reporting an output that never got set.
type RequiredKey = Option<Option<String>>;

/// Deserialize a [`RequiredKey`], recording that the key was present.
///
/// The outer `Option` has to be filled in here rather than left to serde:
/// serde reads a JSON `null` as `None` at every level of nesting, so a plain
/// `Option<Option<String>>` collapses `null` and absent back into the one value
/// this type exists to tell apart. serde reaches this function only for a key
/// that was written, which is exactly what the outer `Some` means.
fn present_key<'de, D>(deserializer: D) -> std::result::Result<RequiredKey, D::Error>
where
    D: Deserializer<'de>,
{
    Ok(Some(Option::deserialize(deserializer)?))
}

/// Exactly the two keys read out of the file. Unknown keys are ignored, so a
/// whole `pulumi stack output --json` parses.
#[derive(Deserialize)]
struct DescriptorFile {
    #[serde(default, deserialize_with = "present_key")]
    public_ip: RequiredKey,
    #[serde(default, deserialize_with = "present_key")]
    fqdn: RequiredKey,
    /// The keys this file has that aren't read. Kept only so a failure can
    /// name them: the mistake is nearly always a neighbouring key, and an
    /// operator staring at "missing `public_ip`" wants to see what was there.
    #[serde(flatten)]
    extra: serde_json::Map<String, serde_json::Value>,
}

impl DescriptorFile {
    /// Every key the file actually had, sorted.
    ///
    /// "Had" means the key was written, whatever its value: a `public_ip` set
    /// to `null` is listed, because the point of this list is to show the
    /// operator what is in front of them.
    fn keys(&self) -> Vec<&str> {
        let mut keys: Vec<&str> = self.extra.keys().map(String::as_str).collect();
        if self.public_ip.is_some() {
            keys.push("public_ip");
        }
        if self.fqdn.is_some() {
            keys.push("fqdn");
        }
        keys.sort_unstable();
        keys
    }
}

impl NodeDescriptor {
    /// Read and validate a descriptor, naming the node after the file.
    ///
    /// Reading is the only way to build one from JSON: every failure names the
    /// offending file, which is the whole point of a hand-writable format.
    pub fn load(path: &Path) -> Result<Self> {
        let bytes = std::fs::read(path).map_err(|e| Error::read(path, e))?;
        let name = path
            .file_stem()
            .ok_or_else(|| {
                Error::gate(format!(
                    "descriptor path {} has no file name",
                    path.display()
                ))
            })?
            .to_string_lossy()
            .into_owned();
        Self::parse(name, path, &bytes)
    }

    fn parse(name: String, path: &Path, bytes: &[u8]) -> Result<Self> {
        let file: DescriptorFile =
            serde_json::from_slice(bytes).map_err(|e| Error::json(path, e))?;

        // An absent key, an explicit `null`, and an empty string all fail the
        // same way — a descriptor that can't reach a node — and all three are
        // caught here rather than at the first request. They are reported
        // apart, though, because the fix differs: only the absent case is
        // helped by listing the keys that *were* there, and only it is likely
        // to be a typo. Saying "missing" for a key the operator can plainly see
        // in the file sends them looking in the wrong place.
        let require = |key: &str, value: &RequiredKey| match value {
            Some(Some(value)) if !value.is_empty() => Ok(value.clone()),
            Some(Some(_)) => Err(Error::gate(format!(
                "descriptor {} has an empty `{key}`",
                path.display(),
            ))),
            Some(None) => Err(Error::gate(format!(
                "descriptor {} has `{key}` set to null \
                 (Pulumi emits null for an output that never got set)",
                path.display(),
            ))),
            None => Err(Error::gate(format!(
                "descriptor {} is missing required key `{key}`; got keys {:?}",
                path.display(),
                file.keys(),
            ))),
        };

        Ok(Self {
            name,
            public_ip: require("public_ip", &file.public_ip)?,
            fqdn: require("fqdn", &file.fqdn)?,
        })
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::*;

    const FULL: &[u8] = br#"{"public_ip": "203.0.113.7", "fqdn": "az-1.seismicdev.net"}"#;

    fn parse(bytes: &[u8]) -> Result<NodeDescriptor> {
        NodeDescriptor::parse(
            "dev-bootstrap-node-1".into(),
            &PathBuf::from("nodes/dev-bootstrap-node-1.json"),
            bytes,
        )
    }

    #[test]
    fn parses_the_two_keys() {
        let d = parse(FULL).unwrap();

        assert_eq!(d.name, "dev-bootstrap-node-1");
        assert_eq!(d.public_ip, "203.0.113.7");
        assert_eq!(d.fqdn, "az-1.seismicdev.net");
    }

    #[test]
    fn ignores_extra_keys_so_a_whole_stack_output_parses() {
        let stack_output = br#"{
            "public_ip": "203.0.113.7",
            "fqdn": "az-1.seismicdev.net",
            "resource_group": "dev-bootstrap-node-1",
            "vm_id": "/subscriptions/0000/resourceGroups/dev-bootstrap-node-1"
        }"#;

        assert_eq!(parse(stack_output).unwrap(), parse(FULL).unwrap());
    }

    /// Absent, explicitly `null`, and empty all fail: none reaches a node.
    ///
    /// Each is reported as the mistake it is. Asserting on the distinguishing
    /// phrase, not just on `"public_ip"`, is the point — the key name alone
    /// appears in all three, so a test that checks only that would pass even if
    /// every case reported "missing", which is what an operator staring at a
    /// `public_ip` line in their own file must not be told.
    #[test]
    fn reports_an_absent_null_and_empty_key_apart() {
        let cases = [
            (
                &br#"{"fqdn": "az-1.seismicdev.net"}"#[..],
                "is missing required key `public_ip`",
            ),
            (
                &br#"{"public_ip": null, "fqdn": "az-1.seismicdev.net"}"#[..],
                "has `public_ip` set to null",
            ),
            (
                &br#"{"public_ip": "", "fqdn": "az-1.seismicdev.net"}"#[..],
                "has an empty `public_ip`",
            ),
        ];

        for (bytes, expected) in cases {
            let err = parse(bytes).unwrap_err().to_string();
            assert!(err.contains(expected), "{err}");
            assert!(err.contains("dev-bootstrap-node-1.json"), "{err}");
        }
    }

    /// A key that is present but unusable is never called "missing", and the
    /// key list — which exists to catch a typo — never contradicts the
    /// sentence in front of it by listing the key it just called absent.
    #[test]
    fn a_present_but_unusable_key_is_not_called_missing() {
        for bytes in [
            &br#"{"public_ip": null, "fqdn": "az-1.seismicdev.net"}"#[..],
            &br#"{"public_ip": "", "fqdn": "az-1.seismicdev.net"}"#[..],
        ] {
            let err = parse(bytes).unwrap_err().to_string();
            assert!(!err.contains("missing"), "{err}");
            assert!(!err.contains("got keys"), "{err}");
        }
    }

    /// The failure names the keys that *were* there, because the mistake is
    /// nearly always a neighbouring key rather than an empty file.
    #[test]
    fn a_missing_key_reports_the_keys_the_file_had() {
        let err = parse(br#"{"publicIp": "203.0.113.7", "fqdn": "az-1.seismicdev.net"}"#)
            .unwrap_err()
            .to_string();

        assert!(err.contains("missing required key `public_ip`"), "{err}");
        assert!(err.contains(r#"["fqdn", "publicIp"]"#), "{err}");
    }

    #[test]
    fn names_the_node_after_the_file_and_names_the_file_on_failure() {
        let dir = tempfile::tempdir().unwrap();

        let good = dir.path().join("dev-bootstrap-node-2.json");
        std::fs::write(&good, FULL).unwrap();
        assert_eq!(
            NodeDescriptor::load(&good).unwrap().name,
            "dev-bootstrap-node-2"
        );

        let malformed = dir.path().join("dev-bootstrap-node-3.json");
        std::fs::write(&malformed, b"not json").unwrap();
        let err = NodeDescriptor::load(&malformed).unwrap_err().to_string();
        assert!(err.contains("dev-bootstrap-node-3.json"), "{err}");

        let err = NodeDescriptor::load(&dir.path().join("absent.json"))
            .unwrap_err()
            .to_string();
        assert!(err.contains("absent.json"), "{err}");
    }
}
